librad/src/meta/project.rs

```diff
@@ -184,4 +414,231 @@ pub mod tests {
+        let d3 = prj.canonical_data().unwrap();
+        assert_eq!(&d0, &d1);
+        assert_eq!(&d0, &d2);
+        assert_eq!(&d0, &d3);
+
+        // Check signatures collection contents
+        assert_eq!(3, prj.signatures.len());
+        assert_eq!(1, prj.signatures.iter().filter(|s| s.key == m0).count());
+        assert_eq!(1, prj.signatures.iter().filter(|s| s.key == m1).count());
+        assert_eq!(1, prj.signatures.iter().filter(|s| s.key == m2).count());
+
+        // Check signature verification
+        let data = prj.canonical_data().unwrap();
+        for s in prj.signatures.iter() {
+            let sig = Signature::from_hex_string(&s.sig).unwrap();
+            assert_eq!(sig.verify(&data, s.key.device_key()), true);
+        }
+
+        // Check signature removal
+        prj.remove_signature(&m1);
+        assert_eq!(2, prj.signatures.len());
+    }
+
+    #[test]
+    fn test_project_verification() {
+        let (m0, k0) = new_peer_with_key(42);
+        let (m1, k1) = new_peer_with_key(1);
+        let (m2, k2) = new_peer_with_key(2);
+
+        // A new project is not valid because the owner has not signed it
+        let mut prj = Project::new("foo", &m0);
+        assert!(matches!(prj.check_validity(), Err(Error::SignatureMissing)));
+        assert!(!prj.is_valid());
+        // Adding the signature fixes it
+        prj.add_signature(&k0).unwrap();
+        assert!(matches!(prj.check_validity(), Ok(())));
+        assert!(prj.is_valid());
+        // Adding maintainers without signatures invalidates it
+        prj.add_maintainer(&m1);
+        prj.add_maintainer(&m2);
+        assert!(matches!(prj.check_validity(), Err(Error::SignatureMissing)));
+        // Adding the missing signatures does not fix it: m0 signed a previous revision
+        prj.add_signature(&k1).unwrap();
+        prj.add_signature(&k2).unwrap();
+        assert!(matches!(
+            prj.check_validity(),
+            Err(Error::SignatureVerificationFailed)
+        ));
+        // Cannot sign a project twice with the sme key
+        assert!(matches!(
+            prj.add_signature(&k0),
+            Err(Error::SignatureAlreadyPresent)
+        ));
+        // Removing the signature and re adding it fixes the project
+        prj.remove_signature(&m0);
+        prj.add_signature(&k0).unwrap();
+        assert!(prj.is_valid());
+        // Removing a maintainer invalidates it again
+        prj.maintainers.pop();
+        assert!(matches!(
+            prj.check_validity(),
+            Err(Error::SignatureFromNonMaintainer)
+        ));
+    }
+    #[test]
+    fn test_project_update() {
+        let (m0, k0) = new_peer_with_key(42);
+        let (m1, k1) = new_peer_with_key(1);
+        let (m2, k2) = new_peer_with_key(2);
+        let (m3, k3) = new_peer_with_key(3);
+        let (m4, k4) = new_peer_with_key(4);
+
+        // Empty history is invalid
+        let mut history: Vec<Project> = vec![];
+        assert!(matches!(
```

Comment:
I don't see the usefulness of `matches!` here. It would be better to use `assert_eq!` and that way you get the diff of the two computations back.